include/unifex/with_scheduler_affinity.hpp

```diff
@@ -0,0 +1,166 @@
+    static_cast<Sender&&>(sender),
+    unstoppable(schedule(static_cast<Scheduler&&>(scheduler)))))) {
+  return finally(
+      static_cast<Sender&&>(sender),
+      unstoppable(schedule(static_cast<Scheduler&&>(scheduler))));
+}
+
+template <typename Sender, typename Scheduler>
+using wsa_sender_wrapper =
+    typename _wsa_sender_wrapper<Sender, Scheduler>::type;
+
+template <typename Sender, typename Scheduler>
+class _wsa_sender_wrapper<Sender, Scheduler>::type final {
+  using sender_t =
+      decltype(_make_sender(UNIFEX_DECLVAL(Sender), UNIFEX_DECLVAL(Scheduler)));
+
+  sender_t sender_;
+
+public:
+  template <
+      template <typename...>
+      typename Variant,
+      template <typename...>
+      typename Tuple>
+  using value_types = sender_value_types_t<sender_t, Variant, Tuple>;
+
+  template <template <typename...> typename Variant>
+  using error_types = sender_error_types_t<sender_t, Variant>;
+
+  static constexpr bool sends_done = sender_traits<sender_t>::sends_done;
+
+  static constexpr blocking_kind blocking = sender_traits<sender_t>::blocking;
+
+  static constexpr bool is_always_scheduler_affine = true;
+
+  template <typename Sender2, typename Scheduler2>
+  type(Sender2&& sender, Scheduler2 scheduler) noexcept(noexcept(_make_sender(
+      static_cast<Sender2&&>(sender), static_cast<Scheduler2&&>(scheduler))))
+    : sender_(_make_sender(
+          static_cast<Sender2&&>(sender),
+          static_cast<Scheduler2&&>(scheduler))) {}
+
+  template(typename Self, typename Receiver)          //
+      (requires same_as<remove_cvref_t<Self>, type>)  //
+      friend auto tag_invoke(tag_t<connect>, Self&& sender, Receiver&& receiver) noexcept(
+          is_nothrow_connectable_v<member_t<Self, sender_t>, Receiver>) {
+    return connect(
+        static_cast<Self&&>(sender).sender_, static_cast<Receiver&&>(receiver));
+  }
+};
+
+struct _fn final {
+  template(typename Sender, typename Scheduler)  //
+      (requires sender<Sender> AND sender_traits<
+          remove_cvref_t<Sender>>::is_always_scheduler_affine)  //
+      constexpr Sender&&
+      operator()(Sender&& s, Scheduler&&) const noexcept {
+    // this is the identity function for statically-affine senders
+    return static_cast<Sender&&>(s);
+  }
+
+  template(typename Sender, typename Scheduler)                              //
+      (requires sender<Sender> AND scheduler<Scheduler> AND                  //
+       (!sender_traits<remove_cvref_t<Sender>>::is_always_scheduler_affine)  //
+       AND tag_invocable<_fn, Sender, Scheduler>)                            //
+      constexpr auto
+      operator()(Sender&& s, Scheduler&& sched) const
+      noexcept(is_nothrow_tag_invocable_v<_fn, Sender, Scheduler>)
+          -> tag_invoke_result_t<_fn, Sender, Scheduler> {
+    // allow customization
+    return tag_invoke(
+        _fn{}, static_cast<Sender&&>(s), static_cast<Scheduler&&>(sched));
+  }
```

Comment:
I need to change these constraints.  If there's a customization, it should always be invoked; if there *isn't* a customization, then `is_always_scheduler_affine` just changes the default implementation.